### ISSUE TITLE
Empty where blocks should get dead code warnings

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#4623]
```agda
  test : Set₁
  test = Set
    module M where
      -- postulate A : Set -- without this declaration, M is not in scope

  module N = M

  -- No module M in scope
  -- when scope checking the declaration
  --   module N = M
```
Empty where blocks are ignored, even if named.  There should be a warning and deadcode highlighting.  Otherwise, it is surprising that the module name is undefined.

**Comment by guillaumebrunerie** [comment]
Or an empty module should be in scope?

**Comment by andreasabel** [comment]
> Or an empty module should be in scope?

I have opted for the conservative path, warnings for empty `where` blocks.  I think in release code, you would not want empty `where` blocks, and during working on a file, one can ignore the warnings.

This is a continuation of theme #947.

**Comment by gallais** [comment]
You should get a warning for the empty `where` but perhaps the module should
be in scope too? It was after all explicitly named and an out of scope error
is a fatal error so if you start from a valid program

```agda
f : Set
f = ? module M where
  g : Set
  g = ?

module N = M
```

and move the definition of `g` to the toplevel during a refactoring, suddenly
you end up with a fatal error!

**Comment by guillaumebrunerie** [comment]
Or maybe out of scope errors should just be warnings? That’s how it works in many programming languages :D (just kidding of course)

**Comment by gallais** [comment]
> Or maybe out of scope errors should just be warnings? 

There is something to be said for that. We could decide to replace the offending
name with a hole (containing the out of scope one) and see whether we can keep
processing the file in a sensible manner.

Similarly we could have a block-aware parser and happily process the parts of
the file that do parse (if you have a parse error in a RHS, it's perfectly
sensible to postulate the function that was declared & keep going for instance).

**Comment by andreasabel** [comment]
@gallais, while you are right, there has been a conspicuous absence of complaints about this situation, which has been around since the fix of #947.  
I conclude that the pattern in the OP and your comment does not exist "in the wild" (only in-vitro).
Thus, it is likely irrelevant whether the `M` is in scope or not.

My patch does not change the logic that already existed in Agda, only adds the warning.  My original motivation for this fix was to clarify the existing situation, such that I can establish the invariant that lists of `where`-declarations in the `Abstract` syntax are non-empty always.

If people feel strongly about having empty where-module names in scope, I will abstain and leave the door open for that change.

**Comment by andreasabel** [comment]
> Similarly we could have a block-aware parser and happily process the parts of
> the file that do parse

I think this would be the way forward and very much in the spirit of interactive program construction.

**Comment by andreasabel** [comment]
Btw, I fixed this issue to celebrate #3939!  
Working on concrete syntax was such a nuisance before, now there is big penalty any more to modify the modules early in compilation.

**Comment by guillaumebrunerie** [comment]
Just to clarify, I personally don’t really care either way :)
